### ISSUE TITLE
Support for @JsonValue

### ISSUE THREAD

**Issue by Quurks** [immutables/immutables#634]
(Immutables 2.5.3 / Jackson 2.8.7)

I am trying to use `@JsonValue` to serialize/deserialize simple wrapper types.

The serialization part works fine, ValueString serializes to a simple String `"alpha"`, ValueList becomse a simple array `["beta"]`.

On deserialization Jackson expects a `@JsonCreator` annotation accepting the serialized data - those are missing in the generated Immutable classes. ImmutableValueString deserializes just fine however - Jackson uses the private constructor (bypassing `@Check` checks). ImmutableValueList deserialization fails with a JsonMappingException because Jackson can't map the array data to the private constructor.

Is it possible to generate the corresponding `@JsonCreator` static factory method when `@JsonValue` is encountered? This would allow it to work in all instances while preserving `@Check` safeguards.

``` java
public class SerializeTest {
    @Immutable
    @JsonDeserialize
    @JsonSerialize
    public abstract static class AbstractValueString {
        @JsonValue public abstract String data();
    }

    @Immutable
    @JsonDeserialize
    @JsonSerialize
    public abstract static class AbstractValueList {
        @JsonValue public abstract List<String> data();
    }

    public static void main(String[] args) throws Exception {
        ObjectMapper mapper = new ObjectMapper(new JsonFactory()).registerModule(new GuavaModule());

        ImmutableValueString valueString = ImmutableValueString.builder().data("alpha").build();
        //This works fine, but bypasses @Check checks.
        System.out.println(
           mapper.readValue(mapper.writeValueAsString(valueString), ImmutableValueString.class));


        ImmutableValueList valueList = ImmutableValueList.builder().addData("beta").build();
        // JsonMappingException: Can not deserialize instance of ImmutableValueList$Json out of START_ARRAY token
        System.out.println(
             mapper.readValue(mapper.writeValueAsString(valueList), ImmutableValueList.class));
    }
}


```

**Comment by elucash** [comment]
Sorry for delay, forgot to look deeper into this. Please, make sure that `JsonDeserialize` references immutable implementation class: `@JsonDeserialize(as = ImmutableValueString.class)`. This seems to be the issue, otherwise `@JsonCreator` would be generated.

**Comment by Quurks** [comment]
Its created in with and without as=... , unfortunately the generated one is useless for @JsonValue.
@JsonValue basically flattens the structure by one level (Object containing a single list is serialized as the single list, without the wrapping object) and thus needs a corresponding @JsonCreator to be read again. 

The autogenerated JsonCreator expects a list wrapped in an object, and thus fails on deseralization.

**Comment by elucash** [comment]
simple `@JsonValue` support implemented in 2.5.6 thank you!

**Comment by before** [comment]
@elucash with this change the *value* module requires JDK 8, because `java.util.Optional` is imported. Is this by intention or an accident?
